coders/qoi.c

```diff
@@ -0,0 +1,450 @@
+  else
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  if (IssRGBCompatibleColorspace(image->colorspace) == MagickFalse)
+   (void) TransformImageColorspace(image,sRGBColorspace,exception);
+  if (IsRGBColorspace(image->colorspace))
+     colorspace=QOI_LINEAR;
+  else
+     colorspace=QOI_SRGB;
+  /*
+    Write QOI header.
+  */
+  (void) WriteBlobString(image,"qoif");
+  (void) WriteBlobMSBLong(image,(unsigned int) image->columns);
+  (void) WriteBlobMSBLong(image,(unsigned int) image->rows);
+  (void) WriteBlobByte(image,channels);
+  (void) WriteBlobByte(image,colorspace);
+  /*
+    Initialize encoding state.
+  */
+  px.rgba.r=0;
+  px.rgba.g=0;
+  px.rgba.b=0;
+  px.rgba.a=255;
+  memset(lut,0,sizeof(lut));
+  run=0;
+  p=GetVirtualPixels(image,0,0,image->columns,image->rows,exception);
+  if (p == (const Quantum *) NULL)
+    return(MagickFalse);
+  /*
+    Do the actual encoding.
+  */
+  end = image->rows * image->columns;
+  for (i=0; i < end; i++) {
+    pp=px;
+    px.rgba.r=ScaleQuantumToChar(GetPixelRed(image,p));
+    px.rgba.g=ScaleQuantumToChar(GetPixelGreen(image,p));
+    px.rgba.b=ScaleQuantumToChar(GetPixelBlue(image,p));
+    if (channels == 4)
+      px.rgba.a=ScaleQuantumToChar(GetPixelAlpha(image,p));
+    p+=GetPixelChannels(image);
+
+    if (pp.v == px.v) {
+      run++;
+      if (run == 62) {
+        (void) WriteBlobByte(image,QOI_OP_RUN | (run - 1));
+        run=0;
+      }
+      continue;
+    }
+    if (run > 0) {
+      (void) WriteBlobByte(image,QOI_OP_RUN | (run - 1));
+      run=0;
+    }
+    idx=QOI_COLOR_HASH(px) % 64;
+    if (lut[idx].v == px.v) {
+      (void) WriteBlobByte(image,QOI_OP_INDEX | idx);
+      continue;
+    }
+    lut[QOI_COLOR_HASH(px) % 64]=px;
+    if (pp.rgba.a == px.rgba.a) {
+      signed char
+        vr,
+        vg,
+        vb,
+        vg_r,
+        vg_b;
+      vr=px.rgba.r - pp.rgba.r;
+      vg=px.rgba.g - pp.rgba.g;
+      vb=px.rgba.b - pp.rgba.b;
+      vg_r=vr - vg;
+      vg_b=vb - vg;
+
+      if (vr > -3 && vr < 2 &&
+          vg > -3 && vg < 2 &&
+          vb > -3 && vb < 2) {
+        (void) WriteBlobByte(image,QOI_OP_DIFF
+                 | (vr + 2) << 4 | (vg + 2) << 2 | (vb + 2));
+      } else if (vg_r >  -9 && vg_r < 8 &&
+                 vg   > -33 && vg   < 32 &&
+                 vg_b >  -9 && vg_b < 2) {
+        (void) WriteBlobByte(image,QOI_OP_LUMA | (vg + 32));
+        (void) WriteBlobByte(image,(vg_r + 8) << 4 | (vg_b +  8));
+      } else {
+        (void) WriteBlobByte(image,QOI_OP_RGB);
+        (void) WriteBlobByte(image,px.rgba.r);
+        (void) WriteBlobByte(image,px.rgba.g);
+        (void) WriteBlobByte(image,px.rgba.b);
+      }
+    } else {
+      (void) WriteBlobByte(image,QOI_OP_RGBA);
+      (void) WriteBlobByte(image,px.rgba.r);
+      (void) WriteBlobByte(image,px.rgba.g);
+      (void) WriteBlobByte(image,px.rgba.b);
+      (void) WriteBlobByte(image,px.rgba.a);
+    }
+  }
+  return(MagickTrue);
```

Comment:
A call to `(void) CloseBlob(image);` is missing before the return and this will keep the file.

Reply:
done